README.md
# JSON Schema Utilities Extension

This extension will provide the following 3 menu `Ctrl + Shift + P` options

- JSON Utils: Validate JSON Schema
- JSON Utils: Generate Mock JSON
- JSON Utils: Generate JSON Schema
- JSON Utils: Validate JSON against JSON schema

## JSON Utils: Validate JSON Schema

This option when selected will try to valide the JSON schema (Active in the text editor) against following json-schema definations

> <http://json-schema.org/draft-04/schema>#
> <http://json-schema.org/draft-06/schema>#

In case of an invalid schema, the errors will be logged into a file, `validateSchema.errors.json` and will be saved in to the same location as the JSON schema.

## JSON Utils: Generate Mock JSON

This option when selected will create a mock JSON from JSON schema (active in the text editor)
The mock json file, `generatedJson.mock.json`, will be saved at the same location as the JSON schema.

## JSON Utils: Generate JSON Schema

This option when selected will create JSON schema (active in the text editor)
The mock json file, `generatedSchema.json`, will be saved at the same location as the JSON file.

## JSON Utils: Validate JSON against JSON Schema

For this action to work, the JSON schema file and JSON file should be open in the split text editor, and active. When this option is selected the JSON (right text editor) will be validated against the JSON schema (left text editor).

If errors, an error file, `validateJson.errors.json` will be saved in the same location as JSON file.

CHANGELOG.md
# Change Log

All notable changes to the "jsonschemautils" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/test/suite/extension.test.ts
import * as assert from 'assert';

// You can import and use all API from the 'vscode' module
// as well as import your extension to test it
import * as vscode from 'vscode';
import * as myExtention from '../../extension';

suite('JSON Schema Validator Test Suite', () => {
	vscode.window.showInformationMessage('Start all tests.');

	test('JSON schema validation succeeded', () => {
		assert.equal(-1, [1, 2, 3].indexOf(5));
		assert.equal(-1, [1, 2, 3].indexOf(0));
	});
});


src/extension.ts
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from "vscode";
import generateJson from "./generate-json";
import validateSchema from "./validate-schema";
import validateJsonAgainstSchema from "./validate-json";
import generateJsonSchema from "./generate-schema";

// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
export function activate(context: vscode.ExtensionContext) {
  // Use the console to output diagnostic information (console.log) and errors (console.error)
  // This line of code will only be executed once when your extension is activated
  console.log(
    'Congratulations, your extension "jsonschemautils" is now active!'
  );

  // The command has been defined in the package.json file
  // Now provide the implementation of the command with registerCommand
  // The commandId parameter must match the command field in package.json
  let schemaValidator = vscode.commands.registerCommand(
    "jsonschemautils.validateSchema",
    validateSchema
  );
  let jsonGenerator = vscode.commands.registerCommand(
    "jsonschemautils.generateJson",
    generateJson
  );
  let jsonValidator = vscode.commands.registerCommand(
    "jsonschemautils.validateJson",
    validateJsonAgainstSchema
  );
  let jsonSchemaGenerator = vscode.commands.registerCommand(
    "jsonschemautils.generateJsonSchema",
    generateJsonSchema
  );

  context.subscriptions.push(schemaValidator, jsonGenerator, jsonValidator, jsonSchemaGenerator);
}

// this method is called when your extension is deactivated
export function deactivate() {}


src/generate-json.ts
import * as jsf from "json-schema-faker";
import * as vscode from "vscode";
import { writeFile } from "./common";

export default async function generateJson() {
  if (!vscode.workspace.workspaceFolders) {
    vscode.window.showInformationMessage("No folder or workspace opened");
    return;
  }

  const editor = vscode.window.activeTextEditor;
  if (!editor) {
    vscode.window.showInformationMessage("No JSON schema file opened");
    return;
  }

  const document = editor.document;
  const jsonSchema = document.getText();
  const faked = jsf.generate(JSON.parse(jsonSchema));
  const fileUri = await writeFile(
    "generatedJson.mock.json",
    JSON.stringify(faked)
  );
  if (fileUri) {
    vscode.window.showTextDocument(fileUri);
  }

  vscode.window.showInformationMessage(`Mock JSON Generated. ${fileUri}`);
}


src/generate-schema.ts
import * as toJsonSchema from "to-json-schema";
import * as vscode from "vscode";
import { writeFile } from "./common";

export default async function generateSchema() {
  if (!vscode.workspace.workspaceFolders) {
    vscode.window.showInformationMessage("No folder or workspace opened");
    return;
  }

  const editor = vscode.window.activeTextEditor;
  if (!editor) {
    vscode.window.showInformationMessage("No JSON file opened");
    return;
  }

  const document = editor.document;
  const json = document.getText();
  const jsonSchema = toJsonSchema(JSON.parse(json));
  const fileUri = await writeFile(
    "generatedSchema.json",
    JSON.stringify(jsonSchema)
  );
  if (fileUri) {
    vscode.window.showTextDocument(fileUri);
  }

  vscode.window.showInformationMessage(`JSON Schema Generated. ${fileUri}`);
}


src/validate-schema.ts
import * as vscode from "vscode";
import * as Ajv from "ajv";
import { posix } from "path";
import { writeFile } from "./common";

export default async function validateSchema() {
  if (!vscode.workspace.workspaceFolders) {
    vscode.window.showInformationMessage("No folder or workspace opened");
    return;
  }

  const editor = vscode.window.activeTextEditor;
  if (!editor) {
    vscode.window.showInformationMessage("No JSON schema file opened");
    return;
  }

  if (editor) {
    const ajv = new Ajv({
      schemaId: "auto",
    });
    ajv.addMetaSchema(require("ajv/lib/refs/json-schema-draft-04.json"));
    ajv.addMetaSchema(require("ajv/lib/refs/json-schema-draft-06.json"));
    const document = editor.document;
    const jsonSchema = document.getText();
    if (ajv.validateSchema(JSON.parse(jsonSchema))) {
      vscode.window.showInformationMessage("JSON schema is valid.");
      return;
    }

    console.log(JSON.stringify(ajv.errors));
    const fileUri = await writeFile(
      "validateSchema.errors.json",
      JSON.stringify(ajv.errors)
    );
    if (fileUri) {
      vscode.window.showTextDocument(fileUri);
    }

    vscode.window.showErrorMessage(`JSON schema is invalid. ${fileUri}`);
    return;
  }
}


src/validate-json.ts
import * as vscode from "vscode";
import * as Ajv from "ajv";
import { posix } from "path";
import { writeFile } from "./common";

export default async function validateJsonAgainstSchema() {
  if (!vscode.workspace.workspaceFolders) {
    vscode.window.showInformationMessage("No folder or workspace opened");
    return;
  }

  const editors = vscode.window.visibleTextEditors;
  if (!editors) {
    vscode.window.showInformationMessage("No JSON schema and JSON file opened");
    return;
  }

  if (editors) {
    const ajv = new Ajv({
      schemaId: "auto",
    });
    ajv.addMetaSchema(require("ajv/lib/refs/json-schema-draft-04.json"));
    ajv.addMetaSchema(require("ajv/lib/refs/json-schema-draft-06.json"));
    const jsonSchemaDoc = editors[0].document;
    const jsonSchema = JSON.parse(jsonSchemaDoc.getText());
    const jsonDoc = editors[1].document;
    const json = JSON.parse(jsonDoc.getText());
    if (ajv.validate(jsonSchema, json)) {
      vscode.window.showInformationMessage(
        "JSON is valid against the JSON schema."
      );
      return;
    }

    console.log(JSON.stringify(ajv.errors));
    const fileUri = await writeFile(
      "validateJson.errors.json",
      JSON.stringify(ajv.errors)
    );
    if (fileUri) {
      vscode.window.showTextDocument(fileUri);
    }

    vscode.window.showErrorMessage(
      `JSON is not valid as per JSON schema. ${fileUri}`
    );
    return;
  }
}
